docs: Consensus Roster Terminology (#9834)
hashgraph/hedera-services: platform-sdk/docs/core/dynamic-address-book/DynamicAddressBook-PlatformDesign.md

# Dynamic Address Book - Platform Design

![Dynamic Address Book Platform Design](DynamicAddressBook-PlatformDesign.drawio.svg)

## Network Connection Effects

Assumption: That the network is fully connected, all nodes connect to each other.

* Need to validate that this is true.

> * Huge gap in testing if not.

Reviewer: I'm not sure this statement should be in the readme

Author: agreed

Author: I'm removing the assumption as well.  I'm leaving the question of partial networking connectivity to the Chatter effort.

### Add Node

Existing nodes need to create or receive mutual TLS gossip connections with the added node.

> This change takes effect when a node is added in the `currentEffectiveRoster` for the `pendingConsensusRound`.

Reviewer 2: I think the wording here can be more clear. The change takes effect when a roster with an added node becomes the `currentEffectiveRoster`.

Author: updating the wording to "This change takes effect when a roster with ___ becomes the currentlyEffectiveRoster...

### Remove Node

Existing nodes need to stop gossiping with the removed node. Events from the removed node can still be gossiped,
validated and come to consensus until the round this new roster takes effect becomes ancient.

> This change takes effect when a node is removed from the `currentEffectiveRoster` for the `pendingConsensusRound`.

Reviewer 2: Same comment as in the "Add Node" section.

### Rotate Signing Key

All nodes need to tear down their existing mutual TLS connection with the node that has had its key rotated and
establish a new connection with the node using the new signing certificate as the trusted root certificate.

This change takes effect when a node's key changes in the `currentEffectiveRoster` for the `pendingConsensusRound`.

### Node Weight Change

This has no effect on the existence of network connections.

## Event Creation Effects

### Add Node

Begin using events from the node as parents of new events.

This change takes effect when a node is added to the `currentEffectiveRoster` for the `pendingConsensusRound`.

> This change will modify consensus weight for the network and the event creation algorithm will be reset.

Reviewer 2: Why does this modify weight for the network? A node can join as zero-weight.

Reviewer 2: What does reseting the algorithm mean?

Author: One could view the addition of a node as an ontological change to the distribution of the weight.  True the total weight of the network is unchanged and the value of weight distributed to nodes is numerically the same as before but because there exists a new node, its a different distribution.  

Mathematically a Function is defined on a Domain.  We have a function mapping weight to nodes.   If the Domain of the function changes, one could argue the new function is a different function.  

@cody-littley, thoughts?

Author: If we are buffering newer rounds than `pendingConsensusRound` and caching the latest tips in each newer round, then when the round is updated and event creation advances to the next `pendingConsensusRound` with its `currentlyEffectiveRoster` the set of tips the event creator is using needs to be updated with the cached tips in the round coming off the buffer.  All the nodes are given the weight distribution in the new `currentlyEffectiveRoster`.   I think this should happen regardless of whether or not the weight actually changed per node.   Safer to reset than to patch the data.

Reviewer 3: There are two steps to resetting the tipset event creator. It's not quite as simple as just clearing everything, but it's also not super complex.

1) for all events being tracked (i.e non-ancient events), update the tipset to include the proper nodes. Each tipset is stored as an array.
2) reset the snapshot to be equal to the tipset of the most recently created event.

Step 1 can be skipped for consensus weight only changes.

We need to take these actions regardless of whether or not a new node has zero stake. This is because the tipset algorithm is willing to use zero stake events as parents.

Need to buffer the latest events in each round. The event creator can build on top of any event that has a

> birth round that is less than or equal to the event creator's current round. Events that have a birth round higher

Reviewer 2: >The event creator can build on top of any event that has a birth round that is less than or equal to the event creator's current round.

Do you mean the event creator's pending round?

Author: yes, I've reorganized the wording, please re-review.

> than the current round need to be buffered before reaching the event creator.

Reviewer 2: ```suggestion
than the pending round need to be buffered before reaching the event creator.
```

Buffer: To ensure that the birth round of the child created is never less than the birth round of the parent,

> Only events that have a birthRound less than or equal to the event creator's current round are let through.

Reviewer 2: ```suggestion
only events that have a `birthRound` less than or equal to the pending consensus round, as known by the event crator, are let through.
```

### Remove Node

> Then the `currentEffectiveRoster` for the `pendingConsensusRound` loses a node, the network weight is modified and

Reviewer: ~~Then~~ *When* the...

Author: updated, thank you.

the event creation algorithm is reset.

### Rotate Signing Key

The `currentEffectiveRoster` contains the key used to sign the event. Rotation of the roster will update the signing
key to use.

### Node Weight Change

Changing network weight in the `currentEffectiveRoster` for the `pendingConsensusRound` should cause the event creation
algorithm to reset.

## Event Validation

For each event, lookup the roster within the `eventHorizon` that matches the `birthRound` of the event. Events
outside of the `eventHorizon` are not validatable.

## Orphan Buffer

> Drop events that have a birthRound outside of the `eventHorizon`.

Reviewer: Is this any different than saying that the orphan buffer discards ancient events?

Reviewer: In that case, I'm not sure it should be the responsibility of the `OrphanBuffer`. I think that would be validated earlier in the pipeline, and allow the `OrphanBuffer` to just assume there aren't any time travelers.

Author: I have added a section with impacts to Gossip which would refuse to receive events outside of the event horizon.

## ISS Detection

Use the effective roster for the round in question.

## Consensus Effects

> uses the `currentEffectiveRoster` for the `pendingConsnensRound`.

Reviewer: Capitalize "uses"

Author: fixed, thank you.

The event creator can build on top of any event that has a birth round that is less than or equal to the event creator's
known value for the `pendingConsensusRound`. Events that have a higher birth round need to be buffered until the event
creator's known `pendingConsensusRound` becomes equal to or higher than their birth round.

Buffer: Only the latest events from each node within the same birthRound need to be buffered.

### Remove Node

When a roster with a removed node becomes the `currentEffectiveRoster` for the `pendingConsensusRound`, the event
creation algorithm is reset.

### Rotate Signing Key

The `currentEffectiveRoster` contains the key used to sign the event. Rotation of the roster will update the signing
key to use for signing self events.

### Node Weight Change

Changing network weight in the `currentEffectiveRoster` for the `pendingConsensusRound` should cause the event creation
algorithm to reset.

## Gossip

### Receiving Events

The `futureEventHorizon` and the `pastEventHorizon` define the thresholds for receiving gossipped Events. Received
events with `birthRound` outside of this node's Event Horizon will be rejected or discarded.

### Sending Events

> A node will store ancient events and send them through gossip to help nodes which are behind.

Reviewer 2: ```suggestion
A node will store ancient, but not expired, events and send them through gossip to help nodes which are behind.
```

Reviewer 2: just a suggestion

### Add Node

When transaction handling creates a roster with a new node in it, it becomes the `greatestEffectiveRoster` for

> the `futureEventHorizon` and this node will be able to receive these events.

Author: @cody-littley, @lpetrovic05,  If the pending consensus round is 100, but we have effective rosters for rounds 101 ... 105, can consensus handle receiving events from a node that is in 101 or 105, but not in 100?

> the `futureEventHorizon` and this node will be able to receive these events.

Reviewer 4: Consensus can handle it, those events would just have a weight of zero. We should discuss with Leemon if there is a reason to disallow this

### Remove Node

When a node is removed from a roster and all rosters containing the node have become ancient, this node will no longer
receive events from the removed node. If received, they will be discarded.

### Rotate Signing Key

This roster change has no effect on whether or not send or receive events through Gossip.

### Node Weight Change

This roster change has no effect on whether or not send or receive events through Gossip.

## Event Validation

For each event, lookup the roster within the `eventHorizon` that matches the `birthRound` of the event. Events
outside the `eventHorizon` are not validatable.

## Orphan Buffer

Drop ancient events that have a birthRound less than `minRoundNonAncient`.

## ISS Detection

Use the effective roster for the round in question.

## Consensus Effects

> Uses the `currentEffectiveRoster` for the `pendingConsnensRound`.

Reviewer 2: ```suggestion
Uses the `currentEffectiveRoster` for the `pendingConsensusRound`.
```

When a round comes to consensus:

* The `latestConsensusRound` and `latestConsensRoster` can be retrieved from the produced `ConsensusRound`.
* A new `NonAncientEventWindow` is produced on an output wire.
* Consensus advances the `pendingConsensusRound` and uses the `currentEffectiveRoster` for that round.
* The `RosterDiff` between the `currentEffectiveRoster` and the `latestConsensusRoster` is produced on an output wire.

## Signature Gathering for State

Use the effective roster for the relevant round when validating signatures on the round's state.

## Reconnect

Validating state signatures.

Use the effective roster for the relevant round when validating signatures on the round's state.

> NOTE: The node must be able to validate 1/2 of the weight with its current effective roster. Since 2/3 of the

Reviewer 2: ```suggestion
NOTE: The node must be able to validate more than 1/2 of the weight with its current effective roster. Since 2/3 of the
```

Reviewer 2: Didn't Leemon say this past Friday that requiring >2/3 signatures to fully sign a state causes problems?

Author: yes, if 1/3 of the network has issues, the state signing would fail to be achieved.

Reviewer 2: I think this is ok for now, but we may need to update this later if we decide the "fully signed" threshold needs to go back to >1/2
